pkg/gatewayserver/gatewayserver.go

```diff
@@ -577,77 +583,70 @@ func (gs *GatewayServer) handleUpstream(conn connectionEntry) {
 
 	hosts := make([]*upstreamHost, 0, len(gs.upstreamHandlers))
 	for name, handler := range gs.upstreamHandlers {
-		handler := handler
-		passDevAddr := func(prefixes []types.DevAddrPrefix, devAddr types.DevAddr) bool {
-			for _, prefix := range prefixes {
-				if devAddr.HasPrefix(prefix) {
-					return true
-				}
-			}
-			return false
-		}
 		host := &upstreamHost{
-			name: name,
-			handler: func(ids *ttnpb.EndDeviceIdentifiers) upstream.Handler {
-				if ids != nil && ids.DevAddr != nil && !passDevAddr(handler.GetDevAddrPrefixes(), *ids.DevAddr) {
-					return nil
-				}
-				return handler
-			},
+			name:     name,
+			handler:  handler,
 			handleCh: make(chan upstreamItem),
 		}
 		hosts = append(hosts, host)
 		defer host.handleWg.Wait()
 	}
 
 	for {
-		ctx := ctx
-		var val interface{}
+		var (
+			ctx = ctx
+			val interface{}
+		)
 		select {
 		case <-ctx.Done():
 			return
 		case msg := <-conn.Up():
 			ctx = events.ContextWithCorrelationID(ctx, fmt.Sprintf("gs:uplink:%s", events.NewCorrelationID()))
 			msg.CorrelationIDs = append(msg.CorrelationIDs, events.CorrelationIDsFromContext(ctx)...)
 			val = msg
+			registerReceiveUplink(ctx, gtw, msg.UplinkMessage, protocol)
 		case msg := <-conn.Status():
 			ctx = events.ContextWithCorrelationID(ctx, fmt.Sprintf("gs:status:%s", events.NewCorrelationID()))
 			val = msg
+			registerReceiveStatus(ctx, gtw, msg, protocol)
 		case msg := <-conn.TxAck():
 			ctx = events.ContextWithCorrelationID(ctx, fmt.Sprintf("gs:tx_ack:%s", events.NewCorrelationID()))
 			msg.CorrelationIDs = append(msg.CorrelationIDs, events.CorrelationIDsFromContext(ctx)...)
 			if msg.Result == ttnpb.TxAcknowledgment_SUCCESS {
-				registerSuccessDownlink(ctx, conn.Gateway())
+				registerSuccessDownlink(ctx, gtw, protocol)
 			} else {
-				registerFailDownlink(ctx, conn.Gateway(), msg)
+				registerFailDownlink(ctx, gtw, msg, protocol)
 			}
 			// TODO: Send Tx acknowledgement upstream (https://github.com/TheThingsNetwork/lorawan-stack/issues/76)
 			continue
 		}
+		item := upstreamItem{ctx, val}
 		for _, host := range hosts {
-			item := upstreamItem{
-				ctx:  ctx,
-				val:  val,
-				host: host,
-			}
+			host := host
 			select {
 			case host.handleCh <- item:
 			default:
 				if atomic.LoadInt32(&host.handlers) < maxUpstreamHandlers {
 					atomic.AddInt32(&host.handlers, 1)
 					host.handleWg.Add(1)
-					go handleFn(host)
-					go func(host *upstreamHost) {
+					registerUpstreamHandlerStart(ctx, host.name)
+					go func() {
+						defer host.handleWg.Done()
+						defer atomic.AddInt32(&host.handlers, -1)
+						defer registerUpstreamHandlerStop(ctx, host.name)
```

Comment:
How do we correlate the start and stop events? These calls are concurrent and there's on per uplink.

Reply:
I don’t understand

What is per uplink? This is per handler

Also see the metric; it’s per host

Reply:
> Also see the metric; it’s per host

Yeah I meant one per uplink per host.

> We could track handler lifetime separately in a metric

I thought that was what this metric was intended to measure and hence the comment.